_posts/2019-10-14-Introducing-Pygeos.md

```diff
@@ -0,0 +1,141 @@
+
+```python
+import pygeos
+import numpy as np
+
+points = [
+    pygeos.Geometry("POINT (1 9)"),
+    pygeos.Geometry("POINT (3 5)"),
+    pygeos.Geometry("POINT (7 6)")
+]
+box = pygeos.box(2, 2, 7, 7)
+pygeos.contains(box, points)
+```
+```
+>>> array([False,  True, False])
+```
+
+Here, we computed whether the three `points` are contained in `box` in one go.
+
+## GEOS (Geometry Engine - Open Source)
+PyGEOS merely wraps [GEOS](https://geos.osgeo.org/) in Python. The core `pygeos.Geometry` object contains a C-pointer to the actual `GEOSGeometry` struct.
+
+```python
+box._ptr
+```
+```
+94625728372472
+```
+
+Working with raw C-pointers has a big advantage: there is no overhead from the Python interpreter when accessing the object. This is why PyGEOS is fast. The flip side is that memory leaks and segfaults are very easy to come by.
+
+### Garbage collecting
+The `pygeos.Geometry` uses the Python internals to be able to clean the underlying `GEOSGeometry` when it is not used anymore. In general, each Python object keeps a *reference count*, which is increased when you make another reference (`box2 = box`), and decreased when the variable goes out of scope.
+
+Whenever the reference count becomes zero, Python (more specifically: the *garbage collector*) automatically calls the deallocation method on `pygeos.Geometry`, which destroys the underlying `GEOSGeometry`. The bottomline is that once this is set up, we do need to worry about memory management anymore.
+
+This was the first time I actually did something with the Python C-API. I recommend reading [the Python docs](https://docs.python.org/3/c-api/typeobj.html) or having a look at the [PyGEOS source](https://github.com/pygeos/pygeos/blob/master/src/pygeom.c) if you're interested in how this works exactly.
+
+## Numpy ufuncs
+The functions in PyGEOS are Numpy *universal functions* or *ufuncs* ([further reading](https://docs.scipy.org/doc/numpy/user/c-info.ufunc-tutorial.html)). This makes them array-aware and able to do all the broadcasting gymnastics you may know from NumPy. An example:
+
+```python
+import pygeos
+import numpy as np
+
+# create a 5 boxes that have increasing x-coordinates
+poly_x = np.array([pygeos.box(0 + i, 0, 10 + i, 10) for i in range(5)])
+# create a 5 boxes that have increasing y-coordinates
+poly_y = np.array([pygeos.box(0, 0 + i, 10, 10 + i) for i in range(5)])
+# intersect all combinations
+squares = pygeos.intersection(poly_x[:, np.newaxis], poly_y[np.newaxis, :])
+# compute the areas
+pygeos.area(squares)
+```
+```
+array([[100.,  90.,  80.,  70.,  60.],
+       [ 90.,  81.,  72.,  63.,  54.],
+       [ 80.,  72.,  64.,  56.,  48.],
+       [ 70.,  63.,  56.,  49.,  42.],
+       [ 60.,  54.,  48.,  42.,  36.]])
+```
+
+## Reprojections using PyPROJ
+PyGEOS provides a way to `apply` functions to coordinates directly. This can be used to translate, scale, or reproject you geometries. An example using PyPROJ to projecting a point from lat/lon to webmercator:
```

Comment:
```suggestion
PyGEOS provides a way to `apply` functions to coordinates directly. This can be used to translate, scale, or reproject you geometries. An example using `pyproj` to projecting a point from lat/lon to webmercator:
```

(I think they typically don't capitalize the name of the python package)